src/asl.ts

```diff
@@ -2850,58 +2866,154 @@ export class ASL {
+    const filterResult = this.newHeapVariable();
+
+    const predicateBody = ASLGraph.joinSubStates(
+      predicate.body,
+      ...predicate.body.statements.map((stmt) =>
+        this.evalStmt(stmt, {
+          End: undefined,
+          ResultPath: predicateResult,
+          Next: ASLGraph.DeferNext,
+        })
+      )
     );
+
+    return {
+      startState: "init",
+      states: {
+        init: {
+          Type: "Pass" as const,
+          Parameters: {
+            // copy array to tail it
+            "arr.$": valueJsonPath.jsonPath,
+            arrStr: "[null",
+          },
+          // use the eventual result location as a temp working space
+          ResultPath: filterResult,
+          Next: "check",
+        },
+        check: {
+          Type: "Choice",
+          Choices: [
+            { ...ASL.isPresent(`${filterResult}.arr[0]`), Next: "assign" },
+          ],
+          Default: "end",
+        },
+        // assign the item parameter the head of the temp array
+        assign: ASLGraph.updateDeferredNextStates(
+          { Next: "predicate" },
+          // TODO: support index and array?
+          (predicate.parameters.length > 0
+            ? this.evalDecl(predicate.parameters[0], {
+                jsonPath: `${filterResult}.arr[0]`,
+              })
+            : undefined) ?? { Type: "Pass" }
+        ),
+        // run the predicate function logic
+        predicate: ASLGraph.updateDeferredNextStates(
+          { Next: "checkPredicate" },
+          predicateBody ?? { Type: "Pass" }
+        ),
+        // check that the predicate value is true
+        checkPredicate: {
+          Type: "Choice",
+          Choices: [
+            { ...ASL.isTruthy(predicateResult), Next: "predicateTrue" },
+          ],
+          Default: "predicateFalse",
+        },
+        // tail and append
+        predicateTrue: {
+          Type: "Pass",
+          Parameters: {
+            "arr.$": `${filterResult}.arr[1:]`,
+            // const arrStr = `${arrStr},${JSON.stringify(arr[0])}`;
+            "arrStr.$": `States.Format('{},{}', ${filterResult}.arrStr, States.JsonToString(${filterResult}.arr[0]))`,
+          },
+          ResultPath: filterResult,
+          Next: "check",
+        },
+        // tail
+        predicateFalse: {
+          Type: "Pass",
+          InputPath: `${filterResult}.arr[1:]`,
+          ResultPath: `${filterResult}.arr`,
+          Next: "check",
+        },
+        // parse the arrStr back to json and assign over the filter result value.
+        end: {
+          startState: "format",
+          states: {
+            format: {
+              Type: "Pass",
+              // filterResult = { result: JSON.parse(filterResult.arrStr + "]") }
+              Parameters: {
+                // cannot use intrinsic functions in InputPath or Result
+                "result.$": `States.StringToJson(States.Format('{}]', ${filterResult}.arrStr))`,
```

Comment:
Where does the leading `[` come from?

Reply:
the `[null` above. Set while also setting the temp array.